File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        
        hare = head
        tortoise = head

        #Find Mid point
        while hare and hare.next:
            hare = hare.next.next
            tortoise = tortoise.next
        
        #Reverse second half
        curr = tortoise
        prev = None

        while curr:
            curr.next ,prev,curr = prev,curr,curr.next
        
        #Find max of twin
        max_val = 0

        start = head

        while prev:

            max_val = max(max_val, prev.val + start.val)
            prev = prev.next
            start = start.next

        return max_val 

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {

        //Hare Tortoise 
        ListNode hare = head;
        ListNode tortoise = head;

        while(hare != null && hare.next != null){
            hare = hare.next.next;
            tortoise = tortoise.next;
        }

        //Reverse LinkedList
        ListNode prev = null;
        ListNode curr = tortoise;

        while(curr != null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        //calculate max_val
        int max_val = 0;
        ListNode start = head;

        while(prev != null){

            max_val = Math.max(max_val, prev.val + start.val);
            prev = prev.next;
            start = start.next;
        }

        return max_val;
        
    }
}